src/snowflake/snowpark/lineage.py

```diff
@@ -0,0 +1,454 @@
+            object_name, object_domain, directions, object_version
+        )
+        response = self._session.sql(query_string)
+        json_response = json.loads(response.collect()[0][0])
+
+        rows = []
+        for direction in directions:
+            edges = (
+                json_response.get(_DGQLFields.DATA, {})
+                .get(_DGQLFields.NODE, {})
+                .get(direction.value, [])
+            )
+
+            for edge in edges:
+                if _DGQLFields.SOURCE in edge and _DGQLFields.TARGET in edge:
+                    rows.append(
+                        (
+                            edge[_DGQLFields.SOURCE],
+                            edge[_DGQLFields.TARGET],
+                            direction,
+                            current_depth,
+                        )
+                    )
+
+        return rows
+
+    def _recursive_trace(
+        self,
+        object_name: str,
+        object_domain: str,
+        direction: str,
+        current_depth: int,
+        total_depth: int,
+        object_version: Optional[str] = None,
+    ) -> List[Tuple[VariantType, VariantType, StringType, int]]:
+        """
+        Recursively traces lineage by making successive DGQL queries based on response nodes.
+        """
+        if current_depth > total_depth:
+            return []
+
+        lineage_edges = self._get_lineage(
+            object_name, object_domain, [direction], object_version, current_depth
+        )
+        if (
+            current_depth == total_depth
+            or not lineage_edges
+            or self._is_terminal_entity(lineage_edges[0][0])
+            or self._is_terminal_entity(lineage_edges[0][1])
+        ):
+            return lineage_edges
+
+        next_object = (
+            lineage_edges[0][1]
+            if direction == LineageDirection.DOWNSTREAM
+            else lineage_edges[0][0]
+        )
+        deeper_lineage_edges = self._recursive_trace(
+            next_object[_ObjectField.NAME],
+            next_object[_ObjectField.DOMAIN],
+            direction,
+            current_depth + 1,
+            total_depth,
+            next_object.get(_ObjectField.VERSION),
+        )
+
+        lineage_edges.extend(deeper_lineage_edges)
+        return lineage_edges
+
+    def _is_terminal_entity(self, entity) -> bool:
+        """
+        Determines if the entity should not be explored further.
+        """
+        return entity[_ObjectField.STATUS] in {"MASKED", "DELETED"}
+
+    def _get_name_and_version(self, graph_entity):
+        """
+        Extracts and returns the name and version from the given graph entity.
+        """
+        if graph_entity[_ObjectField.USER_DOMAIN] in self.versioned_object_domains:
+            if graph_entity[_ObjectField.USER_DOMAIN] == _UserDomain.FEATURE_VIEW:
+                if "$" in graph_entity[_ObjectField.NAME]:
+                    parts = graph_entity[_ObjectField.NAME].split("$")
```

Comment:
the safest way to check it's a feature view is to query the relevant tags on the table. Is there a viable way to do that here?

Reply:
We are doing all those things in the server side and the populate userDomain to FeatureView